### PR TITLE
Core: Fix startup hang caused by watchStorySpecifiers

### LINKED ISSUE #18881: Twice as many FSWatchers created when using storyStoreV7
**Describe the bug**

TL;DR: I don't think we should be passing relative paths to watchpack since it's effectively causing 2x as many calls to `fs.watch`. This isn't a bug per-se, but it's _probably_ not great for performance, and it makes encountering another bug with watchpack on osx more common.

When running with `storyStoreV7` enabled we create a watcher via [watchpack](https://github.com/webpack/watchpack) in [watch-story-specifiers.ts](https://github.com/storybookjs/storybook/blob/c8bcbeb5106108fc5c1c2fbf55999630ae3d2358/code/lib/core-server/src/utils/watch-story-specifiers.ts#L34). For a configuration like this:

```json
["../src/**/*.stories.mdx", "../src/**/*.stories.@(js|jsx|ts|tsx)"]
```

The `NormalizedStoriesSpecifier`s passed to `watchStorySpecifiers` look like this:

```js
[
  {
    titlePrefix: "",
    directory: "./src",
    files: "**/*.stories.mdx",
    importPathMatcher: /^\.[\\/](?:src(?:\/(?!\.)(?:(?:(?!(?:^|\/)\.).)*?)\/|\/|$)(?!\.)(?=.)[^/]*?\.stories\.mdx)$/,
  },
  {
    titlePrefix: "",
    directory: "./src",
    files: "**/*.stories.@(js|jsx|ts|tsx)",
    importPathMatcher: /^\.[\\/](?:src(?:\/(?!\.)(?:(?:(?!(?:^|\/)\.).)*?)\/|\/|$)(?!\.)(?=.)[^/]*?\.stories\.(js|jsx|ts|tsx))$/,
  },
];
```

And ultimately we call `wp.watch({ directories: ["./src"] })`.

Watchpack internally re-uses watchers, with caching keyed off the path being watched. Webpack also internally uses watchpack, but it uses absolute paths instead of relative ones, so the watchers created by storybook can't be shared with those created by webpack.

We can demonstrate this by creating a test project with `storyStoreV7` enabled, adding this to the top of `node_modules/.bin/start-storybook` to log `fs.watch` calls, and then calling `yarn run start-storybook`.

```js
const fs = require("fs");
const originalWatch = fs.watch;
fs.watch = (...args) => {
  console.log(`fs.watch: ${args}`);
  return originalWatch(...args);
};
```

You should see two `fs.watch` calls for all the directories being watched, one relative created by storyook and one absolute created later by webpack after the initial compilation completes. If you change `/node_modules/@storybook/core-server/dist/cjs/utils/watch-story-specifiers.js` to use absolute paths instead by wrapping `ns.directory` in `_path.default.resolve` you'll see half as many.

This is triggering another bug where watchpack can potentially hang for a long time after 2000+ watchers are created on osx. Since all directories are being watched twice, this threshold gets hit a lot sooner.

**System**

```
Environment Info:

  System:
    OS: macOS 12.4
    CPU: (12) x64 Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
  Binaries:
    Node: 16.16.0 - /usr/local/bin/node
    Yarn: 1.22.17 - ~/.yarn/bin/yarn
    npm: 8.11.0 - /usr/local/bin/npm
  Browsers:
    Chrome: 103.0.5060.134
    Safari: 15.5
  npmPackages:
    @storybook/addon-actions: ^6.5.7 => 6.5.7
    @storybook/addon-essentials: ^6.5.7 => 6.5.7
    @storybook/addon-interactions: ^6.5.7 => 6.5.7
    @storybook/addon-links: ^6.5.7 => 6.5.7
    @storybook/builder-webpack5: ^6.5.7 => 6.5.7
    @storybook/jest: ^0.0.10 => 0.0.10
    @storybook/manager-webpack5: ^6.5.7 => 6.5.7
    @storybook/node-logger: ^6.5.7 => 6.5.7
    @storybook/preset-create-react-app: ^4.1.2 => 4.1.2
    @storybook/react: ^6.5.7 => 6.5.7
    @storybook/testing-library: ^0.0.11 => 0.0.11
```

### PR DESCRIPTION
**Pull request by heyimalex**
Closes #18881

## What I did

Changed how watchpack gets initialized in `watchStorySpecifiers`. Previously we:
- Passed in the directories from the story specifiers directly into watchpack as relative paths.

Now we:
- Pass in the directories from the story specifiers _and_ all nested files and directories.
- Pass everything in as absolute paths instead of relative paths.

This fixes two issues:
1. https://github.com/webpack/watchpack/issues/222: Because we passed in only the top level directories, watchpack's recursive watching triggered a lot of churn once you cross the `WATCHPACK_WATCHER_LIMIT` threshold. The outcome was multi-second hangs on startup. See the repro in the linked issue, it's incredibly simple and reliably fails on OSX.
2. #18881: Because we're now passing the same absolute paths to watchpack that webpack passes internally, it's able to internally re-use FSWatcher objects (given the same version of watchpack is being used under the hood).

Reproduce the original issue by checking out `next` and setting up a sandbox. `cd` into the sandbox.

Run `yarn storybook` and note the time to startup. On my machine, 334 ms for manager and 880 ms for preview.

Open node and run this to create a bunch of empty subdirectories within the sandbox.

```js
const fs = require("fs");
const path = require("path");

const base = path.resolve("./src/extra");

// Generate a bunch of directories to simulate watchpack runs against.
function makeDirectoryTree(depth, dirsPerLevel) {
  fs.mkdirSync(base);
  let total = 0;
  let level = [base];
  for (let i = 0; i < depth; i++) {
    const next = [];
    for (const dir of level) {
      for (let j = 0; j < dirsPerLevel; j++) {
        const subdir = path.join(dir, `subdir-${i}-${j}`);
        fs.mkdirSync(subdir);
        total += 1;
        next.push(subdir);
      }
    }
    level = next;
  }
  console.log(`${total} directories created`);
}

makeDirectoryTree(5, 6);
```

Now run `yarn storybook` again. On my machine, **44 s** for manager and 703 ms for preview.

Checkout this branch and run the test again. On my machine, 254 ms for manager and 669 ms for preview.

### Testing

Updated some tests, didn't love how it went.

#### The changes in this PR are covered in the following automated tests:
- [x] stories
- [x] unit tests
- [ ] integration tests
- [ ] end-to-end tests

#### Manual testing

Did `console.log` testing / manual testing on the shape of everything in a sandbox.

### Documentation

N/A

- [ ] Add or update documentation reflecting your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

- [ ] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli/src/sandbox-templates.ts`
- [ ] Make sure this PR contains **one** of the labels below:
   <details>
     <summary>Available labels</summary>

     - `bug`: Internal changes that fixes incorrect behavior.
     - `maintenance`: User-facing maintenance tasks.
     - `dependencies`: Upgrading (sometimes downgrading) dependencies.
     - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
     - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
     - `documentation`: Documentation **only** changes. Will not show up in release changelog.
     - `feature request`: Introducing a new feature.
     - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
     - `other`: Changes that don't fit in the above categories.
   
   </details>

### CODE REVIEW

**Comment by chrisgarber** [comment]
Big savings 👍 🎉

**Comment by yannbf** [comment]
Hey @heyimalex thanks a lot for your PR. So the `storyStoreV7` flag doesn't exist anymore in Storybook 8. I'm assuming you are still facing the same issues in Storybook 8?


@tmeasday could you please take a look at this PR? Thanks!

**Comment by shilman** [comment]
@heyimalex thanks so much for putting this together. i just created [a new PR that adds another Watchpack watcher](https://github.com/storybookjs/storybook/pull/26634). It's watching the config directory (e.g. `.storybook/`) and has slightly different logic than the current set of watchers. i don't know much about the performance implications of my work, but if you could possibly take a look and let me know if you have any concerns, i would really appreciate that!

**Comment by heyimalex** [comment]
@yannbf Sorry, I've updated the title. Yes, this is still an issue, it's been around for as long as `watchStorySpecifiers` has existed, but just now getting around to upstreaming it. We've had a similar patch in place at dayjob for almost two years now, back when `storyStoreV7` was experimental 😅 

@shilman I believe the same issue will exist, but it's IMO less likely to be problematic because `.storybook` doesn't tend to contain that many files. Default watcher limit on osx is ~2k, but that applies to _all_ nested directories beneath the passed directory (if they don't match the `ignore` rules), so most folks will only see this issue when the size of their `src` / story containing dir becomes fairly large / deeply nested.

**Comment by tmeasday** [comment]
This seems great @heyimalex. Some questions:

> Pass in the directories from the story specifiers and all nested files and directories.

I'm not really understanding why given we pass in the same set of directories it doesn't actually end up needing to recurse anyway, but it seems like you have a good handle on this :)

> given the same version of watchpack is being used under the hood

I suppose there is no real way we can guarantee this, is there? But it is good that the perf boost is there in the case that it does happen.

File: code/lib/core-server/src/utils/watch-story-specifiers.ts [hunk 1]
```diff
@@ -15,34 +14,58 @@ const isDirectory = (directory: Path) => {
...
 };
 
-// Watchpack (and path.relative) passes paths either with no leading './' - e.g. `src/Foo.stories.js`,
-// or with a leading `../` (etc), e.g. `../src/Foo.stories.js`.
-// We want to deal in importPaths relative to the working dir, so we normalize
-function toImportPath(relativePath: Path) {
-  return relativePath.startsWith('.') ? relativePath : `./${relativePath}`;
+// Takes an array of absolute paths to directories and synchronously returns
+// absolute paths to all existing files and directories nested within those
+// directories (including the passed parent directories).
```
Comment (tmeasday) [inline code/lib/core-server/src/utils/watch-story-specifiers.ts:19]: Does this need to be sync? If we do it async might it give other parts of our boot up a chance to do work while waiting on FS?

**Review by tmeasday — APPROVED** [review]
Code changes LGTM

**Comment by heyimalex** [comment]
> I'm not really understanding why given we pass in the same set of directories it doesn't actually end up needing to recurse anyway, but it seems like you have a good handle on this :)

Forgive me if this isn't perfectly accurate, haven't really worked on this in a couple of years.

So watchpack has direct watchers and recursive watchers. It has a function called [reducePlan](https://github.com/webpack/watchpack/blob/main/lib/reducePlan.js) which takes a mapping of paths to watch, and intelligently collapses them down, creating recursive watchers to replace multiple direct watchers, so that the total number of watchers is below the limit.

When passing in all directories, this reduce plan runs once, and (given the repro conditions) watchpack will recognize that a single recursive watcher is the best move.

When passing in a single top level directory, essentially what happens is every time a readdir call completes, watchpack calls watch on each of the returned directories, and reducePlan runs again. You can see a "trace" of this behavior (created with hacky console log instrumenting) [here](https://github.com/heyimalex/watchpack/blob/mac-slowness-repro/watchpack-behavior-test-output.txt). It ultimately gets to the same plan (see the final `RecursiveWatcher.ctor: /Users/alex/projects/watchpack/fake-dir-base`), but the process churns through the creation and close of a bunch of individual watchers, as reducePlan is working on an incomplete picture.

> I suppose there is no real way we can guarantee this, is there? But it is good that the perf boost is there in the case that it does happen.

No, there isn't, this is more of a best-case-scenario optimization.

Also, just want to call out another possible solution: using a different library. The actual thing we're using watchpack for seems pretty simple on its face: we need to watch directories and trigger a callback when files that match some pattern nested within those directories are added, removed, or updated. Watchpack was probably seen as an easy enough way to get there, and something we already had a transitive dependency on, but given the level of footgunnery here maybe a different dep is appropriate.

File: code/lib/core-server/src/utils/watch-story-specifiers.ts (hunk 1 above)
Comment (heyimalex) [reply code/lib/core-server/src/utils/watch-story-specifiers.ts:19]: I don't think it _must_ be sync, we'd just create the watcher asynchronously after directory walking completes. IMO performance implications here are not significant tho, even with 10k directories readdirSync calls seem to complete in negligible time.

**Comment by tmeasday** [comment]
> Watchpack was probably seen as an easy enough way to get there, and something we already had a transitive dependency on, but given the level of footgunnery here maybe a different dep is appropriate.

Agree 💯 ! If you have another suggestion I'm all ears!

**Comment by heyimalex** [comment]
After some light research, evaluating file watching libs is really hard hahaha. It seems like [@parcel/watcher](https://www.npmjs.com/package/@parcel/watcher) may be a good choice, as it's what's currently used by vscode. [chokidar](https://github.com/paulmillr/chokidar) is widely deployed, but [vscode moved away from it](https://github.com/microsoft/vscode/issues/210422#issuecomment-2060319301). Many other small libraries that seem to have spotty cross platform support or not be maintained.

**Comment by heyimalex** [comment]
Alternative branch [here](https://github.com/storybookjs/storybook/pull/27081), but this PR is definitely lower risk.

**Comment by tmeasday** [comment]
Thanks @heyimalex. We are in feature freeze right now but I'll aim to get this or the other one merged soon after 8.1 goes out.

**Comment by shilman** [comment]
@heyimalex thanks for your patience on this. we're releasing 8.1 in a day or two and i'm planning to merge this as soon as we get that out so it has some time to bake for 8.2. excellent work!!! 🙌

**Comment by tmeasday** [comment]
@shilman we should consider the other PR that migrates us to Parcel's file watcher.

**Comment by shilman** [comment]
Greate call @tmeasday -- I totally missed that. Let's discuss at the next available opportunity!

**Pull request merged**